### PR TITLE
Checkpointing Testing

### PR DESCRIPTION
**Pull request by cgardens**
## What
* Testing Checkpointing (https://github.com/airbytehq/airbyte/pull/3290) required a lot of extra pieces, so splitting it into its own PR so it's easier to consume

## How
* Create a "test" source and destination. These allow us to just define really test specific functionality freely. This is good for testing important edge cases.
* In this case we use `ExceptionAfterNSource` and `LoggingDestination` so that we could write an acceptance test for checkpointing. The strategy is that the source emits N records and then throws an exception. Then we verify that the state afterwards is still retained.
* This required exposing the state somehow. I chose to throw it in the API, since that seems to be something that we're going to do soon anyway and was pretty straight forward.

## Pre-merge Checklist
* [ ] ~~_Add back pressure test that uses `ThrottledDestination` and `InfiniteFeedSource`_~~(punting this: https://github.com/airbytehq/airbyte/issues/3596)

## Recommended reading order
1. `E2ETestSource.java`
1. `E2ETestDestination.java`
1. `config.yaml`
1. the rest

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1200367912513076/1200368130293484) by [Unito](https://www.unito.io)

### CODE REVIEW

File: airbyte-integrations/connectors/destination-e2e-test/src/main/java/io/airbyte/integrations/destination/e2e_test/E2ETestDestination.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+import io.airbyte.integrations.base.IntegrationRunner;
+import io.airbyte.protocol.models.AirbyteConnectionStatus;
+import io.airbyte.protocol.models.AirbyteMessage;
+import io.airbyte.protocol.models.ConfiguredAirbyteCatalog;
+import java.util.Map;
+import java.util.function.Consumer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class E2ETestDestination extends BaseConnector implements Destination {
```
Comment (michel-tricot) [inline airbyte-integrations/connectors/destination-e2e-test/src/main/java/io/airbyte/integrations/destination/e2e_test/E2ETestDestination.java:41]: Except for the name, this is brilliant!

What about `End2EndTestDestination` or `TestingDestination`?

File: airbyte-integrations/connectors/destination-e2e-test/src/main/resources/spec.json [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+          },
+          "millis_per_record": {
+            "description": "Time to pause in between records.",
+            "type": "integer"
+          }
+        }
+      }
+    ]
+  }
+}
```
Comment (michel-tricot) [inline airbyte-integrations/connectors/destination-e2e-test/src/main/resources/spec.json:40]: NIT: newline

File: airbyte-integrations/connectors/source-e2e-test/src/main/java/io/airbyte/integrations/source/e2e_test/E2ETestSource.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+import io.airbyte.protocol.models.AirbyteMessage;
+import io.airbyte.protocol.models.ConfiguredAirbyteCatalog;
+import java.util.Map;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This source is designed to be a switch statement for our suite of highly-specific test sourcess.
+ */
+public class E2ETestSource extends BaseConnector implements Source {
```
Comment (michel-tricot) [inline airbyte-integrations/connectors/source-e2e-test/src/main/java/io/airbyte/integrations/source/e2e_test/E2ETestSource.java:44]: `TestingSources`

**Review by michel-tricot — COMMENTED** [review]
Terrific! Great idea to encode behaviors in testing sources & destinations

File: airbyte-integrations/connectors/source-e2e-test/src/main/java/io/airbyte/integrations/source/e2e_test/InfiniteFeedSource.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.airbyte.integrations.source.e2e_test;
```
Comment (davinchia) [inline airbyte-integrations/connectors/source-e2e-test/src/main/java/io/airbyte/integrations/source/e2e_test/InfiniteFeedSource.java:25]: haha I wrote the same version in Python!

File: airbyte-integrations/connectors/source-e2e-test/src/test/java/io/airbyte/integrations/source/e2e_test/ExceptionAfterNSourceTest.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    assertEquals(getRecordMessage(3L).getRecord().getData(), read.next().getRecord().getData());
+    assertEquals(getRecordMessage(4L).getRecord().getData(), read.next().getRecord().getData());
+    assertEquals(getRecordMessage(5L).getRecord().getData(), read.next().getRecord().getData());
+    assertEquals(getStateMessage(5L).getState().getData(), read.next().getState().getData());
+    assertEquals(getRecordMessage(6L).getRecord().getData(), read.next().getRecord().getData());
+    assertEquals(getRecordMessage(7L).getRecord().getData(), read.next().getRecord().getData());
+    assertEquals(getRecordMessage(8L).getRecord().getData(), read.next().getRecord().getData());
+    assertEquals(getRecordMessage(9L).getRecord().getData(), read.next().getRecord().getData());
+    assertEquals(getRecordMessage(10L).getRecord().getData(), read.next().getRecord().getData());
+    assertEquals(getStateMessage(10L).getState().getData(), read.next().getState().getData());
```
Comment (davinchia) [inline airbyte-integrations/connectors/source-e2e-test/src/test/java/io/airbyte/integrations/source/e2e_test/ExceptionAfterNSourceTest.java:64]: is it worth checking for an exception here?

**Review by davinchia — APPROVED** [review]
Looks great! Feel free to merge after dealing with the names.

One clarification: the unused test source & destination are checked in for convenience right? (I'm just making sure I didn't miss any other tests that were using them)

File: airbyte-api/src/main/openapi/config.yaml [hunk 6]
```diff
@@ -841,6 +841,29 @@ paths:
           description: Connection not found
         "422":
           $ref: "#/components/responses/InvalidInput"
+  /v1/connections/get_state:
```
[the remark this answers is not in the document]
Comment (ChristopheDuong) [reply airbyte-api/src/main/openapi/config.yaml:844]: I think they asked about `/v1/connections/set_state:` or `/v1/connections/update_state:` too, right?

**Review by ChristopheDuong — APPROVED** [review]

**Review by tuliren — APPROVED** [review]

File: airbyte-integrations/connectors/destination-e2e-test/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+
+WORKDIR /airbyte
+
+ENV APPLICATION destination-e2e-test
+
+COPY build/distributions/${APPLICATION}*.tar ${APPLICATION}.tar
+
+RUN tar xf ${APPLICATION}.tar --strip-components=1
+
+LABEL io.airbyte.version=0.3.1
```
[the remark this answers is not in the document]
Comment (cgardens) [reply airbyte-integrations/connectors/destination-e2e-test/Dockerfile:11]: good call. what if we go with dev here actually? not planning to publish this one.

File: airbyte-integrations/connectors/destination-e2e-test/build.gradle [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+    testImplementation project(':airbyte-test-utils')
+
+    testImplementation "org.testcontainers:postgresql:1.15.1"
+
+    integrationTestJavaImplementation project(':airbyte-integrations:bases:standard-destination-test')
+
+    integrationTestJavaImplementation "org.testcontainers:postgresql:1.15.1"
+
+    implementation files(project(':airbyte-integrations:bases:base-java').airbyteDocker.outputs)
```
[the remark this answers is not in the document]
Comment (cgardens) [reply airbyte-integrations/connectors/destination-e2e-test/build.gradle:25]: why? the dockerfile still depends on that base image, no?

File: airbyte-integrations/connectors/source-e2e-test/build.gradle [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+    testImplementation testFixtures(project(':airbyte-integrations:connectors:source-jdbc'))
+    testImplementation project(":airbyte-json-validation")
+    testImplementation project(':airbyte-test-utils')
+
+    testImplementation 'org.testcontainers:postgresql:1.15.1'
+
+    integrationTestJavaImplementation project(':airbyte-integrations:bases:standard-source-test')
+
+    implementation files(project(':airbyte-integrations:bases:base-java').airbyteDocker.outputs)
```
[the remark this answers is not in the document]
Comment (cgardens) [reply airbyte-integrations/connectors/source-e2e-test/build.gradle:31]: same.

File: airbyte-integrations/connectors/source-e2e-test/src/test/java/io/airbyte/integrations/source/e2e_test/ExceptionAfterNSourceTest.java (hunk 5 above)
Comment (cgardens) [reply airbyte-integrations/connectors/source-e2e-test/src/test/java/io/airbyte/integrations/source/e2e_test/ExceptionAfterNSourceTest.java:64]: yes. good call!

File: airbyte-api/src/main/openapi/config.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (cgardens) [reply airbyte-api/src/main/openapi/config.yaml:844]: definitely! i'm looking for directionally correct here. I know this doesn't fully hit the edit state use case.

[the remark this answers is not in the document]
Comment (michel-tricot) [reply airbyte-api/src/main/openapi/config.yaml:844]: shoudln't it be:
`/v1/connections/state/get`
or 
`/v1/state/get`

**Review by michel-tricot — APPROVED** [review]

**Pull request merged**